rat/nuts.py

```diff
@@ -0,0 +1,381 @@
+        self.waiting_threads = numpy.zeros(maximum_threads, dtype=bool)
+        self.context_lock = threading.Lock()
+        self.gradient_computed_condition = threading.Condition(self.context_lock)
+
+        self.arguments = numpy.zeros((maximum_threads, size), dtype=numpy.float32)
+        self.potential_results = numpy.zeros(maximum_threads, dtype=numpy.float32)
+        self.gradient_results = numpy.zeros((maximum_threads, size), dtype=numpy.float32)
+
+        self.metrics = defaultdict(lambda: [0, 0.0])
+        self.maximum_threads = maximum_threads
+        self.local_identity = {}
+
+    def _single_value_and_grad(self, q0: numpy.array) -> Tuple[float, numpy.array]:
+        device_U, device_gradient = self.value_and_grad_negative_log_density(q0)
+        return float(device_U), numpy.array(device_gradient)
+
+    def _get_ident(self):
+        threading_identity = threading.get_ident()
+
+        try:
+            return self.local_identity[threading_identity]
+        except:
+            if len(self.local_identity) == self.maximum_threads:
+                raise Exception("This potential is already at it's maximum number of threads")
+
+            new_local_identity = len(self.local_identity)
+
+            self.local_identity[threading_identity] = new_local_identity
+
+            return new_local_identity
+
+    def _value_and_grad(self, q0: numpy.array) -> Tuple[float, numpy.array]:
+        active_thread = self._get_ident()
+
+        if not self.active_threads[active_thread]:
+            raise Exception("Called gradient without activating thread")
+
+        self.arguments[active_thread, :] = q0
+
+        while True:
+            with self.context_lock:
+                self.waiting_threads[active_thread] = True
+
+                number_waiting_threads = self.waiting_threads.sum()
+
+                if (self.active_threads.sum() - number_waiting_threads) == 0:
+                    start = time.time()
+
+                    device_U, device_gradients = self.value_and_grad_negative_log_density(self.arguments[self.waiting_threads, :])
+
+                    self.potential_results[self.waiting_threads] = device_U
+                    self.gradient_results[self.waiting_threads, :] = device_gradients
+
+                    self.gradient_computed_condition.notify_all()
+
+                    metrics = self.metrics[number_waiting_threads]
+
+                    metrics[0] += number_waiting_threads
+                    metrics[1] = 0.1 * (time.time() - start) / number_waiting_threads + 0.9 * metrics[1]
+
+                    self.waiting_threads[active_thread] = False
+
+                    break
+
+            # There's a race condition here that hopefully doesn't lead to errors. If a non-running thread doesn't
+            # get to wait before the running thread sets and clears the event, it will wait for the next round to run
+            with self.context_lock:
```

Comment:
For every step, record the values where you want to calculate the gradient values for. And then a single thread calculates the gradients for each chain. Once it's finished it returns the values back to each chain thread. Is this correct or did I understand it wrong?

Reply:
Yeah -- it's a synchronization mechanism where every thread that has signaled it is active (by doing the activate_thread context manager) lines up to have gradients computed, but only one thread makes the actual call.

But I'm about to change this and try something else.

Reply:
I removed the thread synchronization. Now the Potential class is more or less pre-allocating memory to make copies fast.